### PR TITLE
implement spin polarized

### PR DESCRIPTION
**Pull request by jiaxi98**

### CODE REVIEW

File: jrystal/_src/hamiltonian.py [hunk 1]
```diff
@@ -203,20 +205,15 @@ def hamiltonian_matrix(
...
 
   def hamil_k(k, coeff_k):
     k = jnp.reshape(k, [-1, 3])
-    coeff_k = jnp.expand_dims(coeff_k, axis=1)
 
     def efun(u):
-      _coeff = jnp.einsum("i, *n i a b c -> *n a b c", u, coeff_k)
-      _coeff = jnp.expand_dims(_coeff, axis=-4)
+      _coeff = jnp.einsum(
+        "ik, ijkabc -> ijabc", u.reshape(-1, num_bands), coeff_k
```
Comment (nicktianboli) [inline jrystal/_src/hamiltonian.py:214]: ```# u_shape: [2, num_bands] for polarized case [1, num_bands] for unpol. ```

Better follow the convention of the axis:
"si, skixyz -> skxyz"

File: jrystal/_src/hamiltonian.py [hunk 2]
```diff
@@ -227,15 +224,19 @@ def efun(u):
...
 
-    x = jnp.ones(num_bands, dtype=band_coefficient.dtype)
+    if spin_restricted:
+      x = jnp.ones(num_bands, dtype=band_coefficient.dtype)
+    else:
+      x = jnp.ones(2 * num_bands, dtype=band_coefficient.dtype)
     return complex_hessian(efun, x)
 
-  h1 = jax.vmap(hamil_k, in_axes=(0, 1), out_axes=0)(kpts, band_coefficient)
+  h1 = hamil_k(kpts, band_coefficient)
```
Comment (nicktianboli) [inline jrystal/_src/hamiltonian.py:239]: Should revise the docstring and leave a comment here that only single k point is supported in this function.

File: jrystal/calc/calc_band_structure.py [hunk 3]
```diff
@@ -108,14 +108,29 @@ def calc(
...
+      params, crystal.num_electron, k_vec.shape[0],
+      crystal.spin, config.spin_restricted
     )
 
   coeff_ground_state = pw.coeff(params_pw_ground_state, freq_mask)
   occ_ground_state = get_occupation(params_occ_ground_state)
   ground_state_density_grid = pw.density_grid(
     coeff_ground_state, crystal.vol, occ_ground_state
   )
+  if not config.spin_restricted:
```
Comment (nicktianboli) [inline jrystal/calc/calc_band_structure.py:120]: Can use the code I commented above

File: jrystal/_src/energy.py [hunk 4]
```diff
@@ -304,8 +305,18 @@ def total_energy(
   e_kin = kinetic(g_vector_grid, kpts, coefficient, occupation)
   e_ext = external(density_grid_rec, position, charge, g_vector_grid, vol)
   e_har = hartree(density_grid_rec, g_vector_grid, vol, kohn_sham)
-  if xc == 'lda':
-    e_xc = xc_lda(density_grid, vol, kohn_sham)
+  if xc == 'lda_x':
+    if spin_restricted:
+      e_xc = xc_lda(density_grid, vol, kohn_sham)
+    else:
+      o_alpha = jnp.copy(occupation)
```
[the remark this answers is not in the document]
Comment (jiaxi98) [reply jrystal/_src/energy.py]: This method is incorrect, as the occupation information is not used.
